src/julia-parser.scm

```diff
@@ -1154,18 +1154,20 @@
         (parse-atom s))))
 
 (define (parse-def s is-func anon)
-  (let* ((ex  (parse-unary-prefix s))
-         (sig (if (or (and is-func (reserved-word? ex)) (initial-reserved-word? ex))
-                  (error (string "invalid name \"" ex "\""))
-                  (parse-call-chain s ex #f)))
-         (decl-sig
-          (if (and is-func (eq? (peek-token s) '|::|))
-              (begin (take-token s)
-                     `(|::| ,sig ,(parse-call s)))
-              sig)))
-    (if (eq? (peek-token s) 'where)
-        (parse-where-chain s decl-sig)
-        decl-sig)))
+  (let ((ex  (parse-unary-prefix s)))
+    (if (and (pair? ex) (eq? (car ex) 'macrocall))
+      ex
```

Comment:
This is the only tricky bit --- do we allow `function @mac(x, y)::T` (or with `where`) or does the macro generate the entire signature? Probably the latter (as you have it) but it made me stop and think.

Reply:
I think we can allow both

Reply:
We already have that `@mac(x, y)::T` is allowed in call position and the `::T` is not passed to the macro.

I guess we can allow you to return more syntax than the macro receives, but it seems a bit odd since it means you can end up generating redundant and/or conflicting type assertions.

Reply:
I just mean that the macro can't handle the case that it wants to insert a type assertion when one is already present - the obvious way to do that would have been to pass that syntax to the macro but I don't think we can do that

Unless I'm missing something?